pkg/openstack/redis.go

```diff
@@ -0,0 +1,188 @@
+package openstack
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	redisv1 "github.com/openstack-k8s-operators/infra-operator/apis/redis/v1beta1"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	corev1beta1 "github.com/openstack-k8s-operators/openstack-operator/apis/core/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type redisStatus int
+
+const (
+	redisFailed   redisStatus = iota
+	redisCreating redisStatus = iota
+	redisReady    redisStatus = iota
+)
+
+// ReconcileRedis -
+func ReconcileRedis(
+	ctx context.Context,
+	instance *corev1beta1.OpenStackControlPlane,
+	helper *helper.Helper,
+) (ctrl.Result, error) {
+	var failures []string = []string{}
+	var inprogress []string = []string{}
+
+	// NOTE(beagles): This performs quite a bit of processing to purge
+	// even if the controller is meant to be disabled and might set an
+	// error condition somewhere along the way. I suspect this is
+	// actually desirable because it indicates that the cluster is
+	// somehow in a state that the operator doesn't know how to
+	// reconcile. The implementation will need to careful about the
+	// disabled from the start scenario when it likely should *not*
+	// start appearing in conditions. It also means that the situation
+	// where sample template data in a config where redis is disabled
+	// will require require that the controller iterator over the
+	// specs and make sure that existing deployed redis instances are
+	// deleted. Generally means if we want the redis controller to
+	// truly remain invisible it's best not to provide any example
+	// templates info while the service is disabled.
+
+	// We first remove redises no longer owned
+	redises := &redisv1.RedisList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(instance.Namespace),
+	}
+	if err := helper.GetClient().List(ctx, redises, listOpts...); err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			corev1beta1.OpenStackControlPlaneRedisReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			corev1beta1.OpenStackControlPlaneRedisReadyErrorMessage,
+			err))
+		return ctrl.Result{}, err
+	}
+
+	for _, redis := range redises.Items {
+		for _, ref := range redis.GetOwnerReferences() {
+			// Check owner UID to ensure the redis instance is owned by this OpenStackControlPlane instance
+			if ref.UID == instance.GetUID() {
+				owned := false
+
+				// Check whether the name appears in spec
+				for name := range instance.Spec.Redis.Templates {
+					if name == redis.GetName() {
+						owned = true
+						break
+					}
+				}
+
+				// The instance name is no longer part of the spec. Let's delete it
```

Comment:
As far as I see this will delete Redis instances from the namespace that was created manually and never owned or managed by the openstack operator. That feels wrong.

Reply:
That's a good point. I pulled this bit of code from the memcached. It might be worth having a look to see if that needs modified there.